src/libopensc/card-dnie.c

```diff
@@ -729,6 +729,62 @@ int dnie_match_card(struct sc_card *card)
 	LOG_FUNC_RETURN(card->ctx, result);
 }
 
+static int dnie_sm_free_wrapped_apdu(struct sc_card *card,
+		struct sc_apdu *plain, struct sc_apdu **sm_apdu)
+{
+	struct sc_context *ctx = card->ctx;
+	int rv = SC_SUCCESS;
+
+	LOG_FUNC_CALLED(ctx);
+	if (!sm_apdu)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_INVALID_ARGUMENTS);
+	if (!(*sm_apdu))
+		LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+
+	plain->resp = (*sm_apdu)->resp;
+	plain->resplen = (*sm_apdu)->resplen;
+	plain->sw1 = (*sm_apdu)->sw1;
+	plain->sw2 = (*sm_apdu)->sw2;
```

Comment:
I can spend a few more days in the disentangling and see how far I get.
It would be good if @miguel-cv and @rickyepoderi please take a look at this code in any case. The enveloping has been removed, we need to check if that will cause failures for bigger input data, and in that case, we need to think how to make this enveloping work with the wrapping.